### PR TITLE
Add `HTMLQuerySelectorAll` & `HTMLQuerySelector` scalar functions

### LINKED ISSUE #1915: [Task]: HTML Scalar Functions
### What needs to be done?

Ideally we should cover all functions from [DOMParentNode](https://www.php.net/manual/en/class.dom-parentnode.php) 

But we can start from: 

- `ScalarFunctionChain::querySelectorAll()` - returns all
- `ScalarFunctionChain::querySelector()` - returns first 

I don't think we need a dedicated DSL functions as those are type of functions that you execute on `ref('html')`

Like all other scalar functions those two would require: 

- Unit tests covering edge cases
- Simple integration tests covering happy path usage on some example pipeline

### PR DESCRIPTION
**Pull request by stloyd**
Resolves: #1915

<h2>Change Log</h2>
<hr /> 
<div id="change-log">
  <h4>Added</h4>
  <ul id="added">
    <li>Add `HTMLQuerySelectorAll` & `HTMLQuerySelector` scalar functions</li>
  </ul> 
  <h4>Fixed</h4>  
  <ul id="fixed">
    
  </ul>
  <h4>Changed</h4>
  <ul id="changed">
    <li>`HTMLType & `HTMLEntry` require now `\Dom\HTMLDocument`</li>
  </ul>  
  <h4>Removed</h4>
  <ul id="removed">
    
  </ul>
  <h4>Deprecated</h4>
  <ul id="deprecated">
    
  </ul>  
  <h4>Security</h4> 
  <ul id="security">
    
  </ul>     
</div>

### CODE REVIEW

**Comment by stloyd** [comment]
Coverage is a lie, as it runs on 8.2, but only part of the test is run on 8.4.

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    public function eval(Row $row) : mixed
+    {
+        $value = (new Parameter($this->value))->as($row, type_html());
+        $path = (new Parameter($this->path))->asString($row);
+
+        if (null === $value || null === $path) {
+            return null;
+        }
+
+        if (!\class_exists('\Dom\HTMLDocument')) {
```
Comment (norberttech) [inline src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php]: I think it would be safer to simply throw an exception here since it's not even about the mistake and fault tolerance, without this class query selector can't be even used, which means this function works only on PHP 8.4+

**Comment by norberttech** [comment]
Both scalar functions must come with integration tests (like all other scalar functions) that shows how this works with a full pipeline

File: src/core/etl/tests/Flow/ETL/Tests/Integration/Function/HTMLQuerySelectorAllTest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        self::assertCount(0, $results);
+    }
+
+    #[RequiresPhp('>= 8.4')]
+    public function test_valid_query_all_on_html_document() : void
+    {
+        $rows = df()
+            ->read(from_rows(rows(row(html_entry('html_raw', $this->html)))))
+            ->withEntry('html', ref('html_raw')->htmlQuerySelectorAll('body div span'))
+            ->drop('html_raw')
```
Comment (norberttech) [inline src/core/etl/tests/Flow/ETL/Tests/Integration/Function/HTMLQuerySelectorAllTest.php:40]: what's the type of entry `html` here?

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            throw new \RuntimeException('This function requires \Dom\HTMLDocument extension available in PHP 8.4+.');
+        }
+
+        return HTMLDocument::createFromString($value->toString())
+            ->querySelector($path);
```
Comment (norberttech) [inline src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php]: also now when I think about, what's the purpose of having HTMLEntry if we anyway work on strings? :| We could have literally just use `string_entry`

I think we might be better with adding querySelector and querySelectorAll functions to HTMLDocument and make him to keep HTML as `\DOM\HTMLDocument` for `php >= 8.4` and as string for those below 8.4. 

Shit it's getting a bit complicated at this point due to 8.3 :|

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+{
+    public function __construct(
+        private readonly mixed $value,
+        private readonly ScalarFunction|string $selector,
+    ) {
+    }
+
+    public function eval(Row $row) : ?Element
+    {
+        if (\PHP_VERSION_ID < 80400) {
```
Comment (norberttech) [inline src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php]: the reasony why we always check if specific function/class exists instead of using hardcoded version constraints, is because when you check class/function polyfills can enable this. When you hardcode that it requires 8.4, even with polyfill this won't work. 

Please move towards checking if specific class/function/method exists every where

File: src/lib/types/src/Flow/Types/Type/Logical/HTMLType.php [hunk 5]
```diff
@@ -24,24 +34,38 @@ public function assert(mixed $value) : HTMLDocument
...
+
+        if ('' === $value) {
+            return false;
+        }
+
+        if ('<' !== $value[0]) {
+            return false;
+        }
+
+        return \preg_match(self::HTML_ALIKE_REGEX, $value) === 1;
```
Comment (norberttech) [inline src/lib/types/src/Flow/Types/Type/Logical/HTMLType.php]: all of the above is incorrect, TypeHTML expects `@implements Type<HTMLDocument>` - this means that string, even looking like html is not a valid HTMLDocument

File: src/lib/types/src/Flow/Types/Type/Native/String/StringTypeNarrower.php [hunk 6]
```diff
@@ -33,71 +31,42 @@ public function narrow(mixed $value) : Type
...
-        }
-
-        if ($this->isInteger()) {
-            return type_integer();
-        }
-
-        if ($this->isTimeZone()) {
-            return type_time_zone();
+        if ($value === '') {
+            return type_string();
```
Comment (norberttech) [inline src/lib/types/src/Flow/Types/Type/Native/String/StringTypeNarrower.php:38]: any reason for that?

File: phpstan.neon [hunk 7]
```diff
@@ -60,7 +60,15 @@ parameters:
...
         - src/core/etl/src/Flow/ETL/Formatter/ASCII/ASCIITable.php
+        - src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php
+        - src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php
+        - src/core/etl/src/Flow/ETL/Row/Entry/HTMLEntry.php
         - src/core/etl/src/Flow/ETL/Sort/ExternalSort/RowsMinHeap.php
+        - src/core/etl/tests/Flow/ETL/Tests/Integration/Function/HTMLQuerySelectorAllTest.php
+        - src/core/etl/tests/Flow/ETL/Tests/Integration/Function/HTMLQuerySelectorTest.php
+        - src/core/etl/tests/Flow/ETL/Tests/Unit/Function/HTMLQuerySelectorAllTest.php
+        - src/core/etl/tests/Flow/ETL/Tests/Unit/Function/HTMLQuerySelectorTest.php
+        - src/core/etl/tests/Flow/ETL/Tests/Unit/Row/Entry/HTMLEntryTest.php
```
Comment (norberttech) [inline phpstan.neon]: that's not great, can't we suppress only specific errors? Those paths are also prevent static analysis from doing other checks on those classes. 

+ if you keep it this way and we drop support for 8.2 and 8.3 we wont get even notified that those "exclusions" are not longer needed. 

When you suppress only specific errors, when there is nothing to suppress phpstan will notify us that given suppression is no longer needed.

File: src/lib/types/src/Flow/Types/Type/Native/String/StringTypeNarrower.php (hunk 6 above)
Comment (stloyd) [reply src/lib/types/src/Flow/Types/Type/Native/String/StringTypeNarrower.php:38]: Every check in this narrower has done this assert internally, and at the end it was returning `type_string()`, but if it's an empty string, we can do that check once at the top.

File: src/lib/types/src/Flow/Types/Type/Logical/HTMLType.php (hunk 5 above)
Comment (stloyd) [reply src/lib/types/src/Flow/Types/Type/Logical/HTMLType.php]: Changed, moved the check to narrower.

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php (hunk 4 above)
Comment (stloyd) [reply src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php]: Adjusted, got confused as there was a mention to enforce version check.

File: phpstan.neon (hunk 7 above)
Comment (stloyd) [reply phpstan.neon]: Maybe better to generate a baseline instead of any exclusion? This way, if anything changes in baseline, we will get note about "no error to ignore" report, WDYT?

Comment (stloyd) [reply phpstan.neon]: Here is a dedicated commit doing that: [`0156e72` (#1960)](https://github.com/flow-php/flow/pull/1960/commits/0156e720a9c9cf6d65e50ae9fa34faeb4bcd5878)

Comment (norberttech) [reply phpstan.neon]: The problem with a baseline is that I can't review it. Inline suppressions are much easier for me to take a look and say if dev is just lazy or if it's a true problem like here.

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    ) {
+    }
+
+    /**
+     * @return null|array<Element>
+     */
+    public function eval(Row $row) : ?array
+    {
+        if (!\class_exists('\Dom\HTMLDocument')) {
+            throw new \RuntimeException('This function requires \Dom\HTMLDocument extension available in PHP 8.4+.');
```
Comment (norberttech) [inline src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php]: Pretty much everywhere in Flow we are using dedicated expcetions from Flow namespace. I believe we even have an exception for missing dependency, in this case I would create PHPVersionException

File: phpstan.neon (hunk 7 above)
Comment (stloyd) [reply phpstan.neon]: Removed tests from exclusion, and left only dedicated classes.

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php (hunk 8 above)
Comment (stloyd) [reply src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php]: Done.

File: phpstan.neon [hunk 9]
```diff
@@ -60,6 +60,9 @@ parameters:
     excludePaths:
         - src/cli/src/Flow/CLI/Command/*
         - src/core/etl/src/Flow/ETL/Formatter/ASCII/ASCIITable.php
+        - src/core/etl/src/Flow/ETL/Function/HTMLQuerySelectorAll.php
+        - src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php
+        - src/core/etl/src/Flow/ETL/Row/Entry/HTMLEntry.php
```
Comment (norberttech) [inline phpstan.neon]: how many suppression those would require in file?
(I'm not home today, so sorry if I'm asking redundant questions, not everything is fully visible from phone 😅)

Comment (stloyd) [reply phpstan.neon]: Few are some not really fixable by comment as they are generics related that are saying given class doesn't exists 🫣

Comment (norberttech) [reply phpstan.neon]: Hmm doesn't phpstan allow to suppress specific issues in config for specific file/regex?

Comment (stloyd) [reply phpstan.neon]: It's called baseline which you don't like 🤣

Comment (norberttech) [reply phpstan.neon]: > It's called baseline which you don't like 🤣

It's called "knowing the tools you are using", more details in [documentation](https://phpstan.org/user-guide/ignoring-errors#ignoring-in-configuration-file) 

Errors related to missing \DOM\HTMLDocument can be suppressed like in the example below. 

```
parameters:
	ignoreErrors:
		-
			messages:
				- '#Call to an undefined method [a-zA-Z0-9\\_]+::doFooFoo\(\)#'
				- '#Call to an undefined method [a-zA-Z0-9\\_]+::doFooBar\(\)#'
			path: other/dir/AnotherFile.php
```

Comment (stloyd) [reply phpstan.neon]: I know that tool and what baseline is (which is a list of such ignored errors, nothing more). Done.

File: phpstan.neon [hunk 10]
```diff
@@ -85,6 +85,61 @@ parameters:
         -
             function: 'dj()'
 
+    ignoreErrors:
+        -
+            message: '#^Cloning object of an unknown class Dom\\HTMLDocument\.$#'
```
Comment (norberttech) [inline phpstan.neon]: is message mandatory? 

How I was thinking about it, is something like that: 

```
        -
            identifier: class.notFound
            path: src/core/etl/src/Flow/ETL/Row/Entry/HTMLEntry.php
```

Since this is the only issue we really want to suppress. 

Then you also do the same for related scalar functions and move those suppressions here so after migrating fully to 8.4 there will be just one place to remove them all

Comment (stloyd) [reply phpstan.neon]: I understood that you want to exclude as much verbose as possible, but if not, then I adjusted to use a generic exclusion on HTML-related classes

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public function __construct(
+        private readonly mixed $value,
+        private readonly ScalarFunction|string $selector,
+    ) {
+    }
+
+    public function eval(Row $row) : ?Element
+    {
+        if (!\class_exists('\Dom\HTMLDocument')) {
```
Comment (norberttech) [inline src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php:17]: could you actually move that to constructor? This way the exception is going to be thrown when you just use that function, otherwise it will crash only when pipeline is evaluated and only if it goes into this code.

**Review by norberttech — APPROVED** [review]

File: src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php (hunk 11 above)
Comment (stloyd) [reply src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php:17]: Here into the construct or more at `ScalarFunctionChain::htmlQuerySelectorAll()`?

Comment (norberttech) [reply src/core/etl/src/Flow/ETL/Function/HTMLQuerySelector.php:17]: I would do it here since ScalarFunctionChain will pretty much call `new HTMLQueryselector` which will trigger exception
